collector/sys_user_summary.go

```diff
@@ -0,0 +1,163 @@
+		[]string{"user"}, nil)
+	sysUserSummaryCurrentMemory = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, sysSchema, "current_memory"),
+		"The current amount of allocated memory for the user",
+		[]string{"user"}, nil)
+	sysUserSummaryTotalMemoryAllocated = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, sysSchema, "total_memory_allocated"),
+		"The total amount of allocated memory for the user",
+		[]string{"user"}, nil)
+)
+
+type ScrapeSysUserSummary struct{}
+
+// Name of the Scraper. Should be unique.
+func (ScrapeSysUserSummary) Name() string {
+	return sysSchema + ".user_summary"
+}
+
+// Help describes the role of the Scraper.
+func (ScrapeSysUserSummary) Help() string {
+	return "Collect per user metrics from sys.x$user_summary. See https://dev.mysql.com/doc/refman/5.7/en/sys-user-summary.html for details"
+}
+
+// Version of MySQL from which scraper is available.
+func (ScrapeSysUserSummary) Version() float64 {
+	return 5.7
+}
+
+// Scrape the information from sys.user_summary, creating a metric for each value of each row, labeled with the user
+func (ScrapeSysUserSummary) Scrape(ctx context.Context, db *sql.DB, ch chan<- prometheus.Metric, logger log.Logger) error {
+
+	userSummaryRows, err := db.QueryContext(ctx, sysUserSummaryQuery)
+	if err != nil {
+		return err
+	}
+	defer userSummaryRows.Close()
+
+	var (
+		user                   string
+		statements             uint64
+		statement_latency      uint64
+		statement_avg_latency  float64
+		table_scans            uint64
+		file_ios               uint64
+		file_io_latency        uint64
+		current_connections    uint64
+		total_connections      uint64
+		unique_hosts           uint64
+		current_memory         uint64
+		total_memory_allocated uint64
+	)
+
+	for userSummaryRows.Next() {
+		err = userSummaryRows.Scan(
+			&user,
+			&statements,
+			&statement_latency,
+			&statement_avg_latency,
+			&table_scans,
+			&file_ios,
+			&file_io_latency,
+			&current_connections,
+			&total_connections,
+			&unique_hosts,
+			&current_memory,
+			&total_memory_allocated,
+		)
+		if err != nil {
+			return err
+		}
+		ch <- prometheus.MustNewConstMetric(sysUserSummaryStatements, prometheus.CounterValue, float64(statements), user)
+		ch <- prometheus.MustNewConstMetric(sysUserSummaryStatementLatency, prometheus.CounterValue, float64(statement_latency), user)
+		ch <- prometheus.MustNewConstMetric(sysUserSummaryStatementAvgLatency, prometheus.GaugeValue, float64(statement_avg_latency), user)
+		ch <- prometheus.MustNewConstMetric(sysUserSummaryTableScans, prometheus.CounterValue, float64(table_scans), user)
+		ch <- prometheus.MustNewConstMetric(sysUserSummaryFileIOs, prometheus.CounterValue, float64(file_ios), user)
+		ch <- prometheus.MustNewConstMetric(sysUserSummaryFileIOLatency, prometheus.CounterValue, float64(file_io_latency), user)
+		ch <- prometheus.MustNewConstMetric(sysUserSummaryCurrentConnections, prometheus.CounterValue, float64(current_connections), user)
```

Comment:
This appears to be a `prometheus.GaugeValue`.

Reply:
Fixed here and also for current memory.